### PR TITLE
feat: selective sync (#3877)

### PR DESCRIPTION
**Pull request by kshamajain99**
Signed-off-by: kshamajain99 <kshamajain99@gmail.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] My build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 

issue: https://github.com/argoproj/argo-cd/issues/3877
gitops-engine changes: https://github.com/argoproj/gitops-engine/pull/213

This feature applies only out-of-sync resources when `ApplyOutOfSync=true` option is provided by user. 

#### Test performed:

Scenario: For an application with 10 deployments and 1 svc, deleting svc and syncing entire application (performing same action)

With `ApplyOutOfSync=true` enabled
```
INFO[0040] sync/terminate complete              application=guestbook duration=507.61639ms syncId=00001-NPWVT
```

Without `ApplyOutOfSync=true` enabled
```
INFO[0114] sync/terminate complete              application=guestbook duration=1.983608856s syncId=00002-orWNj
```

### CODE REVIEW

**Review by mayzhang2000 — COMMENTED** [review]
go.mod needs to be updated once the change in gitops-engine is merged.

File: controller/sync.go [hunk 1]
```diff
@@ -168,6 +177,16 @@ func (m *appStateManager) SyncAppState(app *v1alpha1.Application, state *v1alpha
 		}),
 		sync.WithSyncWaveHook(delayBetweenSyncWaves),
 		sync.WithPruneLast(syncOp.SyncOptions.HasOption("PruneLast=true")),
+		sync.WithResourceModificationChecker(syncOp.SyncOptions.HasOption("ApplyOutOfSync=true"), func(target *unstructured.Unstructured, live *unstructured.Unstructured) bool {
+			diffResult, err := diff.Diff(target, live,
+				diff.WithNormalizer(compareResult.diffNormalizer),
+				diff.IgnoreAggregatedRoles(compareOptions.IgnoreAggregatedRoles))
+			if err != nil {
+				log.Warnf("Could not compare resource: %v", err)
+				return false
```
Comment (mayzhang2000) [inline controller/sync.go:186]: If there is err, should we return true?

Comment (kshamajain99) [reply controller/sync.go:186]: yes

File: controller/sync.go [hunk 2]
```diff
@@ -168,6 +177,16 @@ func (m *appStateManager) SyncAppState(app *v1alpha1.Application, state *v1alpha
 		}),
 		sync.WithSyncWaveHook(delayBetweenSyncWaves),
 		sync.WithPruneLast(syncOp.SyncOptions.HasOption("PruneLast=true")),
+		sync.WithResourceModificationChecker(syncOp.SyncOptions.HasOption("ApplyOutOfSync=true"), func(target *unstructured.Unstructured, live *unstructured.Unstructured) bool {
+			diffResult, err := diff.Diff(target, live,
```
Comment (alexmt) [inline controller/sync.go:181]: Please see suggestion in https://github.com/argoproj/gitops-engine/pull/213#pullrequestreview-578825696


The `diffResults` is available in `CompareAppState` method and can be returned in an additional field of `comparisonResult` struct:
https://github.com/argoproj/argo-cd/blob/3967baf0c95fa2678f499f23a773e19249265d99/controller/state.go#L82

**Review by alexmt — CHANGES_REQUESTED** [review]

**Review by alexmt — APPROVED** [review]
LGTM. Merging after gitops engine reference is updated to https://github.com/argoproj/gitops-engine/pull/213

**Pull request merged**